src/interaction/sessionrecording.cpp

```diff
@@ -0,0 +1,1207 @@
+        _state = sessionState::idle;
+        _cleanupNeeded = true;
+        LINFO("Playback session finished");
+    }
+}
+
+void SessionRecording::stopPlayback() {
+    if (_state == sessionState::playback) {
+        _state = sessionState::idle;
+        _cleanupNeeded = true;
+        LINFO("Session playback stopped");
+    }
+}
+
+void SessionRecording::cleanUpPlayback() {
+    global::navigationHandler.stopPlayback();
+
+    Camera * camera = global::navigationHandler.camera();
+    ghoul_assert(camera != nullptr, "Camera must not be nullptr");
+    Scene * scene = camera->parent()->scene();
+    if (_timeline.size() != 0) {
+        unsigned int prevIdx = _timeline[_idxTimeline_cameraPtrPrev].idxIntoKeyframeTypeArray;
+        global::navigationHandler.setFocusNode(scene->sceneGraphNode(_keyframesCamera[prevIdx].focusNode));
+    }
+    global::scriptScheduler.stopPlayback();
+
+    _playbackFile.close();
+
+    //Clear all timelines and keyframes
+    _timeline.clear();
+    _keyframesCamera.clear();
+    _keyframesTime.clear();
+    _keyframesScript.clear();
+    _idxTimeline_nonCamera = 0;
+    _idxTime = 0;
+    _idxScript = 0;
+    _idxTimeline_cameraPtrNext = 0;
+    _idxTimeline_cameraPtrPrev = 0;
+    _hasHitEndOfCameraKeyframes = false;
+
+    _cleanupNeeded = false;
+}
+
+bool SessionRecording::isDataModeBinary() {
+    if (_recordingDataMode == recordDataMode::binary)
+        return true;
+    else
+        return false;
+}
+
+void SessionRecording::writeToFileBuffer(const double& src) {
+    const size_t writeSize_bytes = sizeof(double);
+    unsigned char const *p = reinterpret_cast<unsigned char const*>(&src);
+    memcpy((_keyframeBuffer + _bufferIndex), p, writeSize_bytes);
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(std::vector<char>& cvec) {
+    const size_t writeSize_bytes = cvec.size() * sizeof(char);
+    memcpy((_keyframeBuffer + _bufferIndex), &cvec[0], writeSize_bytes);
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(const unsigned char c) {
+    const size_t writeSize_bytes = sizeof(char);
+    _keyframeBuffer[_bufferIndex] = c;
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::writeToFileBuffer(const bool b) {
+    const size_t writeSize_bytes = sizeof(char);
+    if( b )
+        _keyframeBuffer[_bufferIndex] = 1;
+    else
+        _keyframeBuffer[_bufferIndex] = 0;
+    _bufferIndex += writeSize_bytes;
+}
+
+void SessionRecording::saveStringToFile(const std::string s) {
+    size_t strLen = s.size();
+    size_t writeSize_bytes = sizeof(size_t);
+
+    _bufferIndex = 0;
+    unsigned char const *p = reinterpret_cast<unsigned char const*>(&strLen);
+    memcpy((_keyframeBuffer + _bufferIndex), p, writeSize_bytes);
+    _bufferIndex += (unsigned int)writeSize_bytes;
+    saveKeyframeToFileBinary(_keyframeBuffer, _bufferIndex);
+
+    _recordFile.write(s.c_str(), s.size());
```

Comment:
This will not write the `\0` character, i think .. not sure if it is a problem or not, though

Reply:
Saved files do not have the \0 character, so no change made here.